### DISCUSSION TITLE
TextValidationBehavior - Styles

### DISCUSSION THREAD

**Discussion by Cywizz** [CommunityToolkit/Maui#385] [Q&A]
Hi,

I have 2 questions regarding the TextValidationBehavior I am testing out.

1. When you set the Invalid/Valid style colors...are these colors suppose to override those specified in the Styles.xaml file? 
The reason I ask, is that I can only see a change once I commented out the Normal and Disabled VisualStates for the entry in the Styles.xaml file. I was hoping that I can still define my styles and let any validation process override as needed.

2. I think there is an issue with the behavior unless I am doing something wrong.
I am using the latest RC2 of MAUI and MCT.
If you create a new project and rig up an Entry control to the TextValidationBehavior, I get weird behaviour.
I set the minlength to 5 and max to 10, for this test.

What I found:
- If you start typing (less than 5 characters) it works as expected. (Red)
- Before you reach 5 characters, deselect the entry control, and then click/select back in. Text is now Green (Incorrect). Hovering over it in windows form, also goes to green.
- If you remove the focus it seems to retain to correct color.

I thought it was something in the Styles.xaml that was causing this, so I commented everything out, but I have the same result.

Is this a bug or have I missed something somewhere.

Thanks,
Cornè

**Comment by Cywizz** [comment]
After testing a bit more, I can confirm that the point number 2 above, does NOT happen on Android and it is working as expected...so issue only with Windows Forms.
Number 1 above is still applicable on Android, and I have to remove the Entry's VisualStates setter before it works.

**Reply by bijington** [reply]
I suspect that the Visual States will no play well with the styling. If you applied the same approach by styling a control and also making use of the Visual States I suspect you would see the same behaviour? 

When you refer to Windows Forms do you mean just running on Windows? So the WinUI3 target?

**Reply by Cywizz** [reply]
Thanks for the reply...
As for your first question, I have added a second Entry control and hooked up the TextChanged event in code behind, checking the length of the text entered and setting the colors accordingly and it works perfectly WITH the visual state active in the Styles.xaml. 
However the Entry with the TextValidationBehaviour still does not work as expected. Could it be a timing issue, where the TextValidationBehaviour sets the colors before the Visual State process does?

Your second question, yes I do mean WinUI3, which does work "better" testing with the TextChanged method above with the exception of a "hovering" style still setting the color back, which I can live with.

**Reply by bijington** [reply]
Looking in to our implementation I can see that the validation behavior uses some internally defined VisualStates and therefore it seems like it is unlikely to play with other visual states. See here https://github.com/CommunityToolkit/Maui/blob/d6492311993b34853ba284db99f2467895b9c917/src/CommunityToolkit.Maui/Behaviors/Validators/ValidationBehavior.shared.cs#L372

I suspect that both bits of behaviour (pun not intended) are related to the same issue/implementation detail.

I have to be honest and say I don't know what the right answer is here but I will try to gather some more information and then try to give an opinion.

**Reply by Cywizz** [reply]
Not a problem. I will be using my own implementation for now, so nothing urgent.

I just wanted to highlight the issue, especially for users that will get the Styles.xaml as part of a new maui project template, and tries to implement this behaviour.

**Reply by bijington** [reply]
Thanks. Out of interest would you be willing to share the details on your implementation in case it helps here?

**Reply by Cywizz** [reply]
All I did, in order to move on, is to set the color of the Entry's text in the code behind based on the length of the string...nothing fancy but it works for now.

**Comment by CliffAgius** [comment]
Closed as answered
